HPB-4006 prevent unsafe request data

Treggats opened the pull request:
# issue HPB-4006

With this PR we make it impossible to use `Request` methods which don't validate the input data.

Treggats committed:
HPB-4006 add test for using unvalidated data outside of a request class

Treggats committed:
HPB-4006 add `FormRequest` from vendor Laravel

Treggats committed:
HPB-4005 add stubs

Treggats committed:
HPB-4005 simplify checking the blacklist

Co-Authored-By: Sander Muller <sander@hihaho.com>

Treggats committed:
HPB-4005 implement the rule

Treggats committed:
HPB-4006 add a controller class stub

RobertBoes commented on src/Rules/ScopeRequestValidateMethods.php:
```diff
@@ -0,0 +1,95 @@
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\ShouldNotHappenException;
+
+final class ScopeRequestValidateMethods implements Rule
```
```suggestion
/**
 * @implements \PHPStan\Rules\Rule<\PhpParser\Node\Expr\MethodCall>
 */
 final class ScopeRequestValidateMethods implements Rule
```
Think this will fix the `Access to an undefined property PhpParser\Node::$name.` phpstan error

Treggats replied:
I feel like an idiot forgetting this 😓  but it only solves part of the errors

SanderMuller commented on src/Rules/ScopeRequestValidateMethods.php:
```diff
@@ -0,0 +1,95 @@
+            return true;
+        }
+
+        return false;
```
```suggestion
        return in_array($methodName, [
            'only',
            'input',
            'get',
            'str',
            'string',
            'integer',
            'boolean',
        ]);
```

SanderMuller commented on tests/Rules/ScopeRequestValidateMethodTest.php:
```diff
@@ -0,0 +1,68 @@
+        );
+        self::assertCount(6, $validErrors);
+
+       $this->analyse([__DIR__ . '/stubs/RequestValidateMethodsInNamespaceStub.php'], [
```
indentation in this file is off

SanderMuller commented on src/Rules/ScopeRequestValidateMethods.php:
```diff
@@ -0,0 +1,95 @@
+    /** @throws ShouldNotHappenException */
+    public function processNode(Node $node, Scope $scope): array
+    {
+        if ($scope->getNamespace() === 'App\\Http\\Request') {
```
```suggestion
        if (str_starts_with($scope->getNamespace(), 'App\\Http\\Request')) {
```

We want to allow subfolders too

Treggats committed:
HPB-4006 also check sub namespace

Co-authored-by: Sander Muller <sander@hihaho.com>

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats commented:
Currently dealing with the following error

> Line   src/Rules/ScopeRequestValidateMethods.php                                                                                                    
> --------------------------------------------------------------------------------------------------------------------------------------------------- 
>  63     Calling PHPStan\Reflection\Php\PhpParameterFromParserNodeReflection::getType() is not covered by backward compatibility promise. The method  
         might change in a minor PHPStan version.                                                                                                     
         💡 If you think it should be covered by backward compatibility promise, open a discussion:                                                   
         💡 https://github.com/phpstan/phpstan/discussions                                                                                            
         💡                                                                                                                                           
         💡 See also:                                                                                                                                 
         💡 https://phpstan.org/developing-extensions/backward-compatibility-promise

Treggats committed:
HPB-4006 update minimum version PHPStan (1.11.0)

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats commented on composer.json:
```diff
@@ -23,7 +26,7 @@
     "require": {
         "php": "^8.2",
-        "phpstan/phpstan": "^1.10.55",
+        "phpstan/phpstan": "^1.11.0",
```
first version where identifiers are used

Treggats commented on composer.json:
```diff
@@ -6,7 +6,10 @@
         "psr-4": {
-            "Hihaho\\PhpstanRules\\": "src/"
+            "Hihaho\\PhpstanRules\\": "src/",
+            "Illuminate\\Foundation\\Http\\": "tests/stubs/Illuminate/Foundation/Http",
```
map stubs to the stubbed namespace

Treggats committed:
HPB-4006 simplify expected error message

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 update test name and separate regular request and form request

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 update method names to better reflect its purpose

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 merge illuminate & form request check

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 check for chained allowed methods

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats commented on tests/stubs/Illuminate/Foundation/Http/FormRequest.php:
```diff
@@ -0,0 +1,47 @@
+use Illuminate\Support\ValidatedInput;
+use Symfony\Component\HttpFoundation\Request;
+
+abstract class FormRequest extends Request
```
minimal dummy for the UserRequest to "function"

RobertBoes replied:
Perhaps we should just do `composer require laravel/framework --dev` (and drop the individual illuminate packages)

Treggats replied:
I was trying to avoid adding a dependency xD This thought has crossed my mind.

RobertBoes replied:
Well, it would be a dev dependency, required to execute tests

Treggats replied:
I'm fine with it. @SanderMuller what do you think?

Treggats replied:
<img width="437" alt="image" src="https://github.com/user-attachments/assets/4f36fc42-a156-4810-b9aa-68e0d86b8651">

Treggats committed:
HPB-4006 move to Validation namespace

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 introduce base class

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 split form request and illuminate request rules

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 update name for the illuminate request rule

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 update name for the form request rule

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 update name blacklist check

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 move name helper to the base class

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats commented on src/Rules/Validation/ScopeFormRequestValidateMethods.php:
```diff
@@ -0,0 +1,76 @@
+            return true;
+        }
+
+        if ($varName === 'request' && $methodName === 'safe') {
```
the `safe()` method is allowed when the previous variable is `$request`.

This does, however, makes the assumption that the parameter name of a Request class is always `$request`. So far in the hihaho codebase that is the case, but it is something to consider

Treggats committed:
HPB-4006 remove redundant guard clause

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats commented on src/Rules/Validation/ScopeFormRequestValidateMethods.php:
```diff
@@ -0,0 +1,72 @@
+            return true;
+        }
+
+        return $varName === 'safe' && $this->isValidateMethod($methodName);
```
a validation method chained of the `safe()` method is allowed

Treggats commented on src/Rules/Validation/ScopeFormRequestValidateMethods.php:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    /** @throws ReflectionException */
+    private function hasFormRequestClass(Scope $scope): bool
```
this loops through the methods is the inspected class and checks if the arguments are an instance/extend from the FormRequest class

Treggats commented on src/Rules/Validation/ScopeIlluminateRequestValidateMethods.php:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    /** @throws ReflectionException */
+    private function hasIlluminateRequestClass(Scope $scope): bool
```
this check is a bit simpler than the FormRequest one. Reason is that we directly have a class we need to check. So no need for extra checks or conversions.

Treggats commented on src/Rules/Validation/ScopeValidationMethods.php:
```diff
@@ -0,0 +1,91 @@
+            ->flatten();
+    }
+
+    protected function nameFrom(MethodCall|Variable $var): string
```
helper to get the name from a Node property

Treggats commented on src/Rules/Validation/ScopeValidationMethods.php:
```diff
@@ -0,0 +1,91 @@
+        return $methods;
+    }
+
+    protected function getMethodParameterClassnames(array $method): Collection
```
this method receives a list of methods, for which it will get the parameters and return the type.

Treggats commented on src/Rules/Validation/ScopeValidationMethods.php:
```diff
@@ -0,0 +1,91 @@
+     * @phpstan-return ReflectionMethod[]
+     * @throws ReflectionException
+     */
+    protected function getClassMethods(Scope $scope): array
```
the `Scope` object contains the PHP class that is being inspected. This method gathers all the methods that are present in that class

Treggats committed:
HPB-4006 remove redundant assertion

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats commented on tests/Rules/Validation/ScopeIlluminateRequestValidateMethodTest.php:
```diff
@@ -0,0 +1,52 @@
+    {
+        /** @var Error[] $errors */
+        $errors = $this->gatherAnalyserErrors([__DIR__ . '/../../stubs/App/Http/Controllers/PeopleControllerStub.php']);
+        $validErrors = array_filter(
```
there are more errors present, but we only want to assert the errors from the rules we are introducing into this PR

Treggats commented on tests/Rules/Validation/ScopeIlluminateRequestValidateMethodTest.php:
```diff
@@ -0,0 +1,52 @@
+        );
+        self::assertCount(6, $validErrors);
+
+        $unsafeRequestDataError = static fn (int $line) => [
```
because of the amount of the repeated error messages, I've extracted it to a simple closure

Treggats commented on tests/Rules/Validation/ScopeIlluminateRequestValidateMethodTest.php:
```diff
@@ -0,0 +1,52 @@
+            $unsafeRequestDataError(18),
+            $unsafeRequestDataError(21),
+            [
+                'No error with identifier method.notFound is reported on line 20.',
```
this extra error is present because a `@phpstan-ignore` is used in the stub

Treggats committed:
HPB-4006 remove outdated class

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats commented on tests/stubs/App/Http/Controllers/PeopleControllerStub.php:
```diff
@@ -0,0 +1,41 @@
+            'children' => [
+                'name' => $request->only(['name']),
+            ],
+            'email' => $request->safe()->email, // @phpstan-ignore method.notFound
```
While the `safe()` method does not exist on the Illuminate Request class, it is allowed. So ignoring the `method.notFound` error, in favour of being able to assert against it.

Treggats committed:
HPB-4006 add support for the "validated()" method

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats commented on tests/Rules/Validation/ScopeFormRequestValidateMethodsTest.php:
```diff
@@ -0,0 +1,30 @@
+    #[Test]
+    public function form_request_class_does_not_use_unvalidated_data_outside_its_namespace(): void
+    {
+        $this->analyse([__DIR__ . '/../../stubs/App/Http/Requests/UserRequest.php'], []);
```
usage of validation methods are allowed within the `App\Http\Requests` namespace

Treggats committed:
HPB-4006 cs & dataset update

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 replace illuminate packages with laravel/framework

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 remove illuminate stub class

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats commented on .github/workflows/tests.yml:
```diff
@@ -31,7 +37,7 @@
         - name: Install dependencies
           run: |
-            composer require "illuminate/support:${{ matrix.laravel }}" --no-interaction --no-update
+            composer require "laravel/framework:${{ matrix.laravel }}" --no-interaction --no-update
```
replacing the illuminate packages with laravel/framework, as we need additional classes and it's easier to have one dependency

Treggats commented on composer.json:
```diff
@@ -6,7 +6,9 @@
         "psr-4": {
-            "Hihaho\\PhpstanRules\\": "src/"
+            "Hihaho\\PhpstanRules\\": "src/",
+            "App\\Http\\Controllers\\": "tests/stubs/App/Http/Controllers",
```
mapping namespaces to the stub file location

Treggats commented on composer.json:
```diff
@@ -22,21 +24,16 @@
-        "php": "^8.2",
-        "phpstan/phpstan": "^1.10.55",
-        "illuminate/support": "^10|^11",
-        "laravel/pint": "^1.13.9"
```
the PHPStan and Laravel Pint packages are used during development. So made more sense to have them in the `require-dev`

RobertBoes replied:
`"illuminate/support": "^10|^11",` would need to stay, as without it you'd be able to install the package in Laravel < 10 (or even, not in a Laravel app)
Could also change it for `illuminate/contracts`, not sure what the best one would be

Treggats committed:
HPB-4006 reorder composer packages

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats commented:
Putting in draft while I work on some internal errors 🙈

Treggats committed:
HPB-4006 extract namespace check

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 move allowed methods to a config

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 simplify mapping test stubs to namespace

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 remove the FormRequest rule

it is handled by the other rule

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 rename rule

it not only covers the Illuminate Request

but also the FormRequest and NovaRequest

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 add "usesValidMethod" method to check for allowed methods

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 update getting the node name

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 solidify retrieving method parameter classnames

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 handle anonymous classes

note; PHPStan throws the following error.
Ignoring because I do not have the time to properly figure this out.

Creating new PHPStan\Node\AnonymousClassNode is not covered by backward compatibility promise.
The class might change in a minor PHPStan version.

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats committed:
HPB-4006 add additional info to the tip

using concatenation because I had issues with formatting the test

Signed-off-by: Tonko Mulder <tonko@tonkomulder.nl>

Treggats commented on extension.neon:
```diff
@@ -1,4 +1,31 @@
+
+parameters:
+	scoperequestvalidatemethods:
+		allowedRequestMethods:
```
default valid validation methods, this can be appended by the local PHPStan config.

```diff
diff --git a/phpstan.neon b/phpstan.neon
index f011012500..ee07b64d03 100644
--- a/phpstan.neon
+++ b/phpstan.neon
@@ -50,6 +50,11 @@ parameters:
     # ignore the following error: Called 'Model::make()' which performs unnecessary work, use 'new Model()'.
     noModelMake: false
 
+    scoperequestvalidatemethods:
+        allowedRequestMethods:
+            - 'safe'
+            - 'validated'
+
     # Optional for having a clickable link to PHPStorm
     editorUrl: 'phpstorm://open?file=%%file%%&line=%%line%%'
 
`

Treggats commented on extension.neon:
```diff
@@ -1,4 +1,31 @@
+parametersSchema:
```
PHPStan requires a schema for the parameters used by the rule

Treggats commented on src/Rules/Validation/ScopeRequestValidateMethods.php:
```diff
@@ -0,0 +1,65 @@
+{
+    public function __construct(array $allowedRequestMethods)
+    {
+        $this->allowedRequestMethods = array_unique($allowedRequestMethods);
```
autowired by the PHPStan config. Currently there is no overriding, only appending to this list

Treggats commented on src/Rules/Validation/ScopeValidationMethods.php:
```diff
@@ -0,0 +1,156 @@
+            try {
+                return (new ReflectionClass($className))->getMethods();
+            } catch (ReflectionException) {
+                // @phpstan-ignore phpstanApi.constructor
```
ignoring because PHPStan complains that it is not backwards compatible, but otherwise we'd get a "not found" for anonymous classes.

SanderMuller commented:
@Treggats thanks this was incorporated in [v3.1.0](https://github.com/hihaho/phpstan-rules/releases/tag/v3.1.0)

Treggats commented:
> @Treggats thanks this was incorporated in [v3.1.0](https://github.com/hihaho/phpstan-rules/releases/tag/v3.1.0)

Nice! I thought I'd seen it in the changelog
